Improve jump arrows drawing

shahril96 opened the pull request:
Previously, some jump arrows might be overlapped or being so close together (which makes it not to look nice).

This PR attempted to fix that, and now all jump arrows fit nicely.  :)

eteran commented:
Does this take into account my refactor that I push a little bit ago?

shahril96 commented:
Yes, I've merged your commits (little while ago) with mine.

eteran commented:
OK, cool. I'll review it in a little bit and give @10110111 a chance to comment too! :-)

Thanks again, this stuff is looking great.

eteran commented on src/widgets/QDisassemblyView.cpp:
```diff
@@ -1287,7 +1287,7 @@ void QDisassemblyView::paintEvent(QPaintEvent *) {
 
 		// find suitable arrow horizontal width
-		for (int i = 0; i < jump_arrow_vec.size(); i++) {
+		for (int i = 0; i < (int) jump_arrow_vec.size(); i++) {
```
Isn't `std::vector<>` indexed by an unsigned type like `size_t` or `std::vector<>::size_type`? It's hard to tell, but if `i` is never compared to a signed value, then we can probably just make it a `size_t` and not need this cast.

Also, even though it's more verbose, please prefer c++ style casts :-).

shahril96 replied:
Ah yes, you're correct. Fixed.

eteran commented on src/widgets/QDisassemblyView.cpp:
```diff
@@ -1356,7 +1356,7 @@ void QDisassemblyView::paintEvent(QPaintEvent *) {
 		process->current_thread()->get_state(&state);
 
-		for (int i = 0; i < jump_arrow_vec.size(); i++) {
+		for (int i = 0; i < (int) jump_arrow_vec.size(); i++) {
```
Same as above

shahril96 replied:
Fixed.

shahril96 commented:
You're welcome, happy to contribute a little bit. :)

10110111 commented:
Some comments on the whole changeset including the previous (merged) one follow.

* Enabling antialiasing results in the vertical separator lines being blurred into two pixels instead of being exactly 1-pixel wide. This is not good. (I'd actually prefer the arrows to also be 1-pixel wide, but that's up to you and @eteran to decide.)
* If a jump targets into the middle of an instruction, the arrow line is drawn to the bottom of the viewport, missing the line containing the target byte.

The following screenshot demonstrates both above mentioned problems:
![Screenshot - 251019 - 00:03:39](https://user-images.githubusercontent.com/6376882/67525270-10a3d380-f6bb-11e9-9ae3-3eb50e2dea36.png)

* The arrows in the following screenshot look confusing:
![Screenshot - 251019 - 00:14:03](https://user-images.githubusercontent.com/6376882/67525965-8197bb00-f6bc-11e9-836a-a6284d4175e4.png)

* The arrows for untaken branches are asymmetric with respect to their path line:
![Screenshot - 251019 - 00:17:10](https://user-images.githubusercontent.com/6376882/67526262-18fd0e00-f6bd-11e9-8878-28d210f1b9bd.png)

* The following display is misleading: I've actually come to this state by stepping from 0xf7fd7817 (skipping the conditional jump), but the red arrow hints as if I have just jumped here from 0xf7fd77fb. Maybe the arrow from the origin shouldn't be painted red.
![Screenshot - 251019 - 00:20:45](https://user-images.githubusercontent.com/6376882/67526511-a17bae80-f6bd-11e9-9e1c-a1d18f8d3c52.png)

10110111 commented on src/widgets/QDisassemblyView.cpp:
```diff
@@ -1285,38 +1286,85 @@ void QDisassemblyView::paintEvent(QPaintEvent *) {
-				arrow_horizontal_length = width_downward_outviewport + font_width_ * 2;
-				width_downward_outviewport = arrow_horizontal_length; // update new width
+				// `j` horizontal width is not good, search next
+				if (is_width_good == false) {
```
Why not `!is_width_good`?

shahril96 replied:
I thought using  "== false" looks more readable in this case. But then, I will fix it. :)

10110111 commented on src/widgets/QDisassemblyView.cpp:
```diff
@@ -1285,38 +1286,85 @@ void QDisassemblyView::paintEvent(QPaintEvent *) {
-		int width_upward_outviewport = 0;
-		int width_downward_outviewport = 0;
+		// find suitable arrow horizontal width
+		for (int i = 0; i < jump_arrow_vec.size(); i++) {
```
I'd prefer the `i`, `j` and `k` variables to be named more descriptively. It's hard to keep in mind what three one-letter variables represent when trying to follow the logic.

shahril96 replied:
I've renamed all these three variables.

10110111 commented on src/widgets/QDisassemblyView.cpp:
```diff
@@ -1285,38 +1286,85 @@ void QDisassemblyView::paintEvent(QPaintEvent *) {
 
-				arrow_horizontal_length = width_upward_outviewport + font_width_ * 2;
-				width_upward_outviewport = arrow_horizontal_length;  // update new width
+					JumpArrow& jump_arrow_prev = jump_arrow_vec[k];
```
This reference should be to `const`: you aren't going to change `jump_arrow_vec[k]`.

shahril96 replied:
Yes, you're correct, I've fixed that.

10110111 commented on src/widgets/QDisassemblyView.cpp:
```diff
@@ -1285,38 +1286,85 @@ void QDisassemblyView::paintEvent(QPaintEvent *) {
-			} else {
+					int jump1_min = std::min(
+						jump_arrow.src_line, 
+						jump_arrow.dst_in_viewport ? jump_arrow.dst_line : (is_dst_upward ? 0 : viewport()->height())
```
Why not make a variable for this long subexpression, which is also repeated in the next `max` call? Same for the `jump2_min` and `jump2_max` min/max call arguments.

shahril96 replied:
Fixed.

10110111 commented on src/widgets/QDisassemblyView.cpp:
```diff
@@ -1285,38 +1286,85 @@ void QDisassemblyView::paintEvent(QPaintEvent *) {
+						jump_arrow_prev.dst_in_viewport ? jump_arrow_prev.dst_line : (is_dst_upward_prev ? 0 : viewport()->height())
+					);
+
+					bool cond1 = jump2_max >= jump1_max && jump2_min >= jump1_max;
```
If I understand this logic correctly, you are checking whether whole jump2 arrow is below jump1 one. This is not obvious until the reader analyzes the expression. So the name of `cond1` should be more descriptive. Same concern for `cond2` in the next line.

shahril96 replied:
I've added comments to clarify what the comparison is doing.

10110111 commented on src/widgets/QDisassemblyView.cpp:
```diff
@@ -1285,38 +1286,85 @@ void QDisassemblyView::paintEvent(QPaintEvent *) {
+
+					bool cond1 = jump2_max >= jump1_max && jump2_min >= jump1_max;
+					bool cond2 = jump2_min <= jump1_min && jump2_max <= jump1_min;
+					bool is_jump_overlap = !(cond1 || cond2);	
```
Maybe `jumps_overlap` instead of `is_jump_overlap`. It would sound a bit more natural.

shahril96 replied:
Fixed.

10110111 commented on src/widgets/QDisassemblyView.cpp:
```diff
@@ -1285,38 +1286,85 @@ void QDisassemblyView::paintEvent(QPaintEvent *) {
+					bool is_jump_overlap = !(cond1 || cond2);	
+
+					if (is_jump_overlap && 
+						j == jump_arrow_prev.horizontal_width) {
```
Maybe avoid splitting this line? It seems short enough when merged (especially compared to the ternary conditionals above).

shahril96 replied:
Fixed.

shahril96 commented:
1)

![image](https://user-images.githubusercontent.com/2675341/67547068-99108b80-f730-11e9-9eac-83f91744e10c.png)

Arrow there is red highlighted because there is a direct jump from 0xf7fd77fd, the reason why I do highlight it is to give a clarification that this jump to this address is always taken.

2)

![image](https://user-images.githubusercontent.com/2675341/67547243-189e5a80-f731-11e9-9247-e2df08d32b75.png)

I've fixed this by adding more horizontal length to one of the jump blocks if it is overlap at the edge.

3)

![image](https://user-images.githubusercontent.com/2675341/67549199-4a65f000-f736-11e9-83fb-b5737c7772ec.png)

Fixed where jump now points to the middle of instruction (see screenshot above)

4)

I have removed anti-aliasing code.
If @eteran has other to say I will add it back.

Thank you for all the comments, really appreciated it. :)

10110111 commented:
> I have removed anti-aliasing code.

I think the arrow tips could benefit from antialiasing. It just has to be enabled in `QPainter` right before that code and disabled afterwards. Just be sure to make them still look symmetric (note that in antialiased case the samples are offset by 0.5 px).

10110111 commented:
I'm OK with these changes now. I was thinking of some pixel-hunting tweaks, but I suppose I'll simply make a follow-up PR when I finalize my ideas.

eteran commented:
@10110111 's review has proven to be very worth having! Thanks to both of you. This PR looks great to me. I'll probably merge later today.
